### PR TITLE
fix: issues #2669 #2281

### PR DESCRIPTION
**Pull request by leodemoura**
@semorrison This PR fixes a high priority issue, but will probably break user-defined functions that manipulate discrimination trees.

### CODE REVIEW

**Comment by leodemoura** [comment]
!bench

File: src/Lean/Meta/Tactic/Simp/Rewrite.lean [hunk 1]
```diff
@@ -142,11 +142,24 @@ def tryTheorem? (e : Expr) (thm : SimpTheorem) (discharge? : Expr → SimpM (Opt
         tryTheoremCore lhs xs bis val type e thm (eNumArgs - lhsNumArgs) discharge?
       else
         return none
+
+/--
+Reture a WHNF configuration for retrieving `[simp]` from the discrimination tree.
```
Comment (semorrison) [inline src/Lean/Meta/Tactic/Simp/Rewrite.lean:147]:
```suggestion
Reture=n a WHNF configuration for retrieving `[simp]` from the discrimination tree.
```

File: src/Lean/Meta/Tactic/Simp/Rewrite.lean [hunk 2]
```diff
@@ -142,11 +142,24 @@ def tryTheorem? (e : Expr) (thm : SimpTheorem) (discharge? : Expr → SimpM (Opt
...
+If user has disabled `zeta` and/or `beta` reduction in the simplifier, we must also
+disabled them when retrieving lemmas from discrimination tree. See issues: #2669 and #2281
```
Comment (semorrison) [inline src/Lean/Meta/Tactic/Simp/Rewrite.lean:149]:
```suggestion
disable them when retrieving lemmas from discrimination tree. See issues: #2669 and #2281
```

File: src/Lean/Meta/UnificationHint.lean [hunk 3]
```diff
@@ -26,8 +26,10 @@ structure UnificationHints where
 instance : ToFormat UnificationHints where
   format h := format h.discrTree
 
+def UnificationHints.config : WhnfCoreConfig := {}
```
Comment (semorrison) [inline src/Lean/Meta/UnificationHint.lean:29]: Is this is right value? Unification hints were previously using `simpleReduce := true`, so I thought we should have `iota := ff` here.

File: src/Lean/Meta/DiscrTree.lean [hunk 4]
```diff
@@ -249,16 +249,16 @@ def hasNoindexAnnotation (e : Expr) : Bool :=
 
 /--
 Reduction procedure for the discrimination tree indexing.
-The parameter `simpleReduce` controls how aggressive the term is reduced.
+The parameter `config` controls how aggressive the term is reduced.
```
Comment (semorrison) [inline src/Lean/Meta/DiscrTree.lean:252]:
```suggestion
The parameter `config` controls how aggressively the term is reduced.
```

File: src/Lean/Meta/UnificationHint.lean (hunk 3 above)
Comment (leodemoura) [reply src/Lean/Meta/UnificationHint.lean:29]: Good catch. I fill fix it.

**Pull request merged**